extensions/windows-event-log/CollectorInitiatedSubscription.cpp

```diff
@@ -0,0 +1,789 @@
+    return false;
+  }
+
+  if (vProperty->Type == EcVarTypeNull) {
+    LOG_SUBSCRIPTION_ERROR("!hArray");
+    return false;
+  }
+
+  EC_OBJECT_ARRAY_PROPERTY_HANDLE hArray = vProperty->PropertyHandleVal;
+  utils::ScopeGuard guard_hArray([hArray]() { EcClose(hArray); });
+
+  DWORD dwEventSourceCount{};
+  if (!EcGetObjectArraySize(hArray, &dwEventSourceCount)) {
+    LOG_SUBSCRIPTION_WINDOWS_ERROR("EcGetObjectArraySize");
+    return false;
+  }
+
+  // Add a new EventSource to the EventSources array object.
+  if (!EcInsertObjectArrayElement(hArray, dwEventSourceCount)) {
+    LOG_SUBSCRIPTION_WINDOWS_ERROR("EcInsertObjectArrayElement");
+    return false;
+  }
+
+  for (auto& prop: std::vector<SubscriptionProperty>{{EcSubscriptionEventSourceAddress, s_sourceAddress.value()}, {EcSubscriptionEventSourceEnabled, true}}) {
+    if (!EcSetObjectArrayProperty(hArray, prop.propId_, dwEventSourceCount, 0, &prop.prop_)) {
+      LOG_SUBSCRIPTION_WINDOWS_ERROR("EcSetObjectArrayProperty id: " + std::to_string(prop.propId_));
+      return false;
+    }
+  }
+
+  if (!EcSaveSubscription(hSubscription, NULL)) {
+    LOG_SUBSCRIPTION_WINDOWS_ERROR("EcSaveSubscription");
+    return false;
+  }
+
+  return true;
+}
+
+bool CollectorInitiatedSubscription::subscribe(const std::shared_ptr<core::ProcessContext> &context)
+{
+  logger_->log_debug("CollectorInitiatedSubscription: maxBufferSize_ %lld", s_maxBufferSize.value());
+
+  provenanceUri_ = "winlog://" + computerName_ + "/" + to_string(s_channel.value()) + "?" + to_string(s_query.value());
+
+  subscriptionHandle_ = EvtSubscribe(
+      NULL,
+      NULL,
+      s_channel.value().c_str(),
+      s_query.value().c_str(),
+      NULL,
+      this,
+      [](EVT_SUBSCRIBE_NOTIFY_ACTION action, PVOID pContext, EVT_HANDLE hEvent) {
+        auto pCollectorInitiatedSubscription = static_cast<CollectorInitiatedSubscription*>(pContext);
+
+        auto& logger = pCollectorInitiatedSubscription->logger_;
+
+        if (action == EvtSubscribeActionError) {
+          if (ERROR_EVT_QUERY_RESULT_STALE == (DWORD)hEvent) {
+            logger->log_error("Received missing event notification. Consider triggering processor more frequently or increasing queue size.");
+          } else {
+            logger->log_error("Received the following Win32 error: %x", hEvent);
+          }
+        } else if (action == EvtSubscribeActionDeliver) {
+          DWORD size = 0;
+          DWORD used = 0;
+          DWORD propertyCount = 0;
+
+          if (!EvtRender(NULL, hEvent, EvtRenderEventXml, size, 0, &used, &propertyCount)) {
+            if (ERROR_INSUFFICIENT_BUFFER == GetLastError()) {
+              if (used > s_maxBufferSize.value()) {
+                logger->log_error("Dropping event %x because it couldn't be rendered within %ll bytes.", hEvent, s_maxBufferSize.value());
+                return 0UL;
+              }
+
+              size = used;
+              std::vector<char> buf(size);
```

Comment:
Is there a reason to use a char buffer here? I think this breaks strict aliasing rules, (accessing wchar_t* through char* would be OK, because char can alias any type, but it is not OK do it the other way).
The official example I found also uses LPWSTR: https://docs.microsoft.com/en-us/windows/desktop/wes/rendering-events, so I think this should be `std::vector<wchar_t> buf(size/2);` (or `std::vector<wchar_t> buf(size/2 + 1)`, if we think the Windows API is really stupid.)